src/custom/pages/Claim/InvestmentFlow/index.tsx

```diff
@@ -49,15 +54,52 @@ function _claimToTokenApproveData(claimType: ClaimType, tokenApproveData: TokenA
   }
 }
 
+function _classifyAndFilterClaimData(claimData: EnhancedUserClaimData[], selected: number[]) {
+  const paid: EnhancedUserClaimData[] = []
+  const free: EnhancedUserClaimData[] = []
+
+  claimData.forEach((claim) => {
+    if (claim.isFree) {
+      free.push(claim)
+    } else if (selected.includes(claim.index)) {
+      paid.push(claim)
+    }
+  })
+  return [free, paid]
+}
+
+function _enhancedUserClaimToClaimWithInvestment(
+  claim: EnhancedUserClaimData,
+  investFlowData: InvestClaim[]
+): ClaimWithInvestmentData {
+  const investmentAmount = claim.isFree
+    ? undefined
+    : investFlowData.find(({ index }) => index === claim.index)?.investedAmount
+
+  return { ...claim, ...calculateInvestmentAmounts(claim, investmentAmount) }
+}
+
 export default function InvestmentFlow({ hasClaims, isAirdropOnly, ...tokenApproveData }: InvestmentFlowProps) {
   const { account } = useActiveWeb3React()
-  const { selected, activeClaimAccount, claimStatus, isInvestFlowActive, investFlowStep } = useClaimState()
+  const { selected, activeClaimAccount, claimStatus, isInvestFlowActive, investFlowStep, investFlowData } =
+    useClaimState()
   const { initInvestFlowData } = useClaimDispatchers()
   const claimData = useUserEnhancedClaimData(activeClaimAccount)
 
-  const selectedClaims = useMemo(() => {
-    return claimData.filter(({ index }) => selected.includes(index))
-  }, [claimData, selected])
+  // Filtering and splitting claims into free and selected paid claims
+  // `selectedClaims` are used on step 1 and 2
+  // `freeClaims` are used on step 2
+  const [freeClaims, selectedClaims] = useMemo(
+    () => _classifyAndFilterClaimData(claimData, selected),
+    [claimData, selected]
+  )
+
+  // Merge all claims together again, with their investment data for step 2
+  const allClaims: ClaimWithInvestmentData[] = useMemo(
+    () =>
+      freeClaims.concat(selectedClaims).map((claim) => _enhancedUserClaimToClaimWithInvestment(claim, investFlowData)),
+    [freeClaims, investFlowData, selectedClaims]
+  )
```

Comment:
you don't need two `useMemos` right? the first one could already return allClaims. I say this cause we don't use `freeClaims` `selectedClaims` anywhere but in this memo

Reply:
I've split because as explained with comments in the code, one section is used for step 1 and 2 while other section is only for step 2.
I don't want to update the step 2 path unnecessarily while on step 1